elastic/integrations
# Packages test analyzer

## What does this PR do?

This introduces a new application which can generate reports about the scope of package testing and the results of those tests.

## Checklist



## Author's Checklist


- [ ] Review code. The main logic is inside `report.py` which may not automatically be expanded by GitHub when reviewing this PR.
- [ ] Test to ensure that `report.py <your_options>` when run with `--output=stdout` produces expected output
- [ ] Test to ensure that `report.py <your_options>` when SMTP options set sends an email with expected output

## How to test this PR locally

1. Ensure that you have a recent version of Python 3. (3.5 or above is recommended.) 
1. Create a virtual environment and install requirements from `requirements-dev.txt` in it (`pip install -r requirements-dev.txt` once the environment is active)
1. Ensure that you have credentials allowing you to access the Elasticsearch server which stores stats for Jenkins data. (Contact Observability Robots if needed.)
1. Ensure that you have a GitHub token which can be used with this application.
1. Run `./report.py`, passing in the correct credentials for Elasticsearch and GitHub and the `--output=stdout` option set. You should see HTML printed directly to your screen. You may also redirect this output to a file and view it in web browser if you wish.
1. Ensure that you have GMail credentials which allow you to send email. You may need to generate an `App Password` for your Google account to do so.
1. Remove the `--output=stdout` option and add the SMTP options including the authentication information. Verify that mail is sent and contains data as-expected.

## Features included and excluded

|Feature|Status|
|--------|------|
|Top 10 least-frequently tested packages|:white_check_mark:|
|Top 10 least-frequently tested packages via pipeline tests|:white_check_mark:|
|Top 10 least-frequently tested packages via system tests|:white_check_mark:|
|Top 10 packages with the most pipeline test failures|:white_check_mark:|
|Top 10 packages with the most system test failures|:white_check_mark:|
|System test report |:construction_worker:|

The system test report will require additional discussion and then follow-up work. I will post a larger thread on the original issue outlining the details.

## Related issues


- Original issue in a private repo. I will link backward from there to this PR.

## Screenshots
<img width="351" alt="Capture d’écran 2021-01-19 à 15 52 26" src="https://user-images.githubusercontent.com/111616/105050721-fabfc280-5a65-11eb-9c3f-25b9e511dfcf.png">
<img width="1001" alt="Capture d’écran 2021-01-19 à 16 59 35" src="https://user-images.githubusercontent.com/111616/105059661-71ad8900-5a6f-11eb-8edf-3e9bd0d6d15b.png">
